packages/babel-core/src/transformation/block-hoist-plugin.js

```diff
@@ -41,23 +39,23 @@ const blockHoistPlugin = {
   visitor: {
     Block: {
       exit({ node }) {
-        let hasChange = false;
-        for (let i = 0; i < node.body.length; i++) {
-          const bodyNode = node.body[i];
-          if (bodyNode?._blockHoist != null) {
-            hasChange = true;
-            break;
-          }
-        }
-        if (!hasChange) return;
-
-        node.body = sortBy(node.body, function (bodyNode) {
-          let priority = bodyNode?._blockHoist;
-          if (priority == null) priority = 1;
-          if (priority === true) priority = 2;
-
-          // Higher priorities should move toward the top.
-          return -1 * priority;
+        if (!node.body.find(node => node?._blockHoist != null)) return;
+        // TODO: Babel 9 (?) - remove stabilityMap once Node >= 12 is guaranteed
+        // Array sorting is not stable in earlier Node releases
+        // See: https://v8.dev/blog/array-sort for details
+        const stabilityMap =
+          process.versions.node.split(".")[0] < 12
+            ? new Map(node.body.map((n, idx) => [n, idx]))
+            : null;
+        const priority = node => {
+          if (node?._blockHoist == null) return -1;
+          if (node?._blockHoist === true) return -2;
+          return -1 * node?._blockHoist;
+        };
+        node.body.sort((a, b) => {
+          const result = priority(a) - priority(b);
+          if (result) return result;
+          if (stabilityMap) return stabilityMap.get(a) - stabilityMap.get(b);
```

Comment:
`return 0` after this?  Not sure what `Array#sort` does with undefined return where `result === 0 && !stabilityMap`.

Reply:
Sounds good - what do you think about `return result`?  (even though it'd be guaranteed to be zero in that case, it links the ownership/origin of that value back to the calculation for anyone reading the code)

Reply:
Seems fine, I don't have a strong opinion but my original suggestion accomplished this in a different way:
```js
const result = priority(a) - priority(b);
if (!result && stabilityMap) {
  return stabilityMap.get(a) - stabilityMap.get(b);
}
return result;
```

Reply:
Ok, yep - I like that there are only two `return` statements there.  I guess I have an aversion to checking two conditions in the `if` block if there's a way we can avoid it.

How about

```javascript
const result = priority(a) - priority(b);
if (stabilityMap) {
  return result || stabilityMap.get(a) - stabilityMap.get(b);
}
return result;
```

Reply:
I have no objection to that.

Reply:
Cool.  Thanks for the discussion :)